// vue-easy-lightbox/src/utils/index.js
/* eslint-disable */
import Vue from 'vue';
export var isServer = Vue.prototype.$isServer;
// TODO: prepare for mobile touch event
export var supportsPassive = false;
if (!isServer) {
    try {
        var options = {};
        Object.defineProperty(options, 'passive', {
            get: function () {
                supportsPassive = true;
            }
        });
        window.addEventListener('test-passive', function () { }, options);
    }
    catch (e) { }
}
export var on = function (target, event, handler, passive) {
    if (passive === void 0) { passive = false; }
    if (!isServer) {
        target.addEventListener(event, handler, supportsPassive ? { capture: false, passive: passive } : false);
    }
};
export var off = function (target, event, handler) {
    if (!isServer) {
        target.removeEventListener(event, handler);
    }
};
var toString = Object.prototype.toString;
export function isArray(arg) {
    return toString.call(arg) === '[object Array]';
}
export var isObject = function (arg) {
    return !!arg && toString.call(arg) === '[object Object]';
};
export var isString = function (arg) {
    return !!arg && toString.call(arg) === '[object String]';
};
export function notEmpty(value) {
    return value !== null && value !== undefined;
}
